fog/ledger/server/tests/router_integration.rs

```diff
@@ -598,3 +614,129 @@ async fn overlapping_stores() {
+            address: free_sockaddr(),
+            block_range: BlockRange::new_from_length(i + 1, blocks_per_shard),
+            stores: stores_config,
+        };
+        shards_config.push(shard);
+    }
+    let config = TestEnvironmentConfig {
+        router_address: free_sockaddr(),
+        router_admin_address: free_sockaddr(),
+        shards: shards_config,
+        rate_limit_period: Some(Duration::from_secs(rate_limit_period_secs)),
+        rate_limit_max: Some(NonZeroU32::new(rate_limit_max).unwrap()),
+        omap_capacity: 1000,
+    };
+
+    let mut blocks_config = vec![];
+    let mut key_index = 0;
+    let num_blocks = 3;
+    let users_per_block = 3;
+    let keys_per_user = 3;
+    for _ in 0..num_blocks {
+        let mut block = HashMap::new();
+        for _ in 0..users_per_block {
+            let account = AccountKey::random_with_fog(&mut rng);
+            let mut keys = vec![];
+            for _ in 0..keys_per_user {
+                keys.push(KeyImage::from(key_index));
+                key_index += 1;
+            }
+            block.insert(account.default_subaddress(), keys);
+        }
+        blocks_config.push(block);
+    }
+
+    let grpc_env = Arc::new(grpcio::EnvBuilder::new().build());
+
+    let mut test_environment = create_env(config, blocks_config, grpc_env, logger.clone());
+
+    // Check that we can get all the key images from the shard
+    let keys_per_block = users_per_block * keys_per_user;
+    let keys: Vec<_> = (0..key_index).map(KeyImage::from).collect();
+    let response = test_environment
+        .router_client
+        .check_key_images(&keys)
+        .await
+        .expect("check_key_images failed");
+    assert_eq!(response.results.len(), key_index as usize);
+    for i in 0..key_index {
+        let key = KeyImage::from(i);
+        assert_eq!(response.results[i as usize].key_image, key);
+        assert_eq!(
+            response.results[i as usize].status(),
+            Ok(Some((i / keys_per_block) + 1))
+        );
+        assert_eq!(
+            response.results[i as usize].timestamp_result_code,
+            TimestampResultCode::TimestampFound as u32
+        );
+    }
+
+    // Trigger the rate limit (hammer as hard as possible for 3 seconds)
+    let start = Instant::now();
+    let mut limited = false;
+    while Instant::now().duration_since(start) < Duration::from_secs(rate_limit_period_secs * 3) {
+        if let Err(Grpc(RpcFailure(e))) =
+            test_environment.router_client.check_key_images(&keys).await
+        {
+            if e.code() == grpcio::RpcStatusCode::RESOURCE_EXHAUSTED {
+                limited = true;
+                break;
+            }
+        }
+    }
+
+    assert!(limited, "failed to trigger the rate limit");
+
+    // Reconnect to clear rate limit condition
+    test_environment.router_client.reconnect();
+
+    // Wait 3s for the rate limit to recover
+    tokio::time::sleep(Duration::from_secs(3)).await;
```

Comment:
You're right that I should remove the sleep.

The behavior you're describing is what I was referring to in the PR description. This isn't what we actually want long-term, but because of how gRPC handles `RESOURCE_EXHAUSTED`, it's all we can do in terms of gRPC-level behavior. To get the behavior we'd actually want, we'd have to return the rate limit results in-band, which would be a more disruptive change. Or we could change our Noise sessions to persist across gRPC sessions but that would also be very invasive and we don't want to do it for other reasons. So, yes, this is bad, but it's acceptable badness in the short term.